app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt

```diff
@@ -4,51 +4,164 @@
+     *
      * @param offset How much to offset the list by
      * @param numberOfItems How many items to fetch
-     * @param onComplete A callback that returns the list of [VisitInfo]
+     * @param onComplete A callback that returns the list of [History]
      */
-    fun getHistory(offset: Long, numberOfItems: Long, onComplete: (List<VisitInfo>) -> Unit)
+    fun getHistory(offset: Long, numberOfItems: Long, onComplete: (List<History>) -> Unit)
 }
 
-// A PagedList DataSource runs on a background thread automatically.
-// If we run this in our own coroutineScope it breaks the PagedList
-fun HistoryStorage.createSynchronousPagedHistoryProvider(): PagedHistoryProvider {
-    return object : PagedHistoryProvider {
-
-        override fun getHistory(
-            offset: Long,
-            numberOfItems: Long,
-            onComplete: (List<VisitInfo>) -> Unit
-        ) {
-            runBlockingIncrement {
-                val history = getVisitsPaginated(
-                    offset,
-                    numberOfItems,
-                    excludeTypes = listOf(
-                        VisitType.NOT_A_VISIT,
-                        VisitType.DOWNLOAD,
-                        VisitType.REDIRECT_TEMPORARY,
-                        VisitType.RELOAD,
-                        VisitType.EMBED,
-                        VisitType.FRAMED_LINK,
-                        VisitType.REDIRECT_PERMANENT
+/**
+ * @param historyStorage
+ */
+class DefaultPagedHistoryProvider(
+    private val historyStorage: PlacesHistoryStorage,
+    private val showHistorySearchGroups: Boolean = FeatureFlags.showHistorySearchGroups,
+) : PagedHistoryProvider {
+
+    private lateinit var historyGroups: List<History.Group>
+
+    @Suppress("LongMethod")
+    override fun getHistory(
+        offset: Long,
+        numberOfItems: Long,
+        onComplete: (List<History>) -> Unit,
+    ) {
+        // A PagedList DataSource runs on a background thread automatically.
+        // If we run this in our own coroutineScope it breaks the PagedList
+        runBlockingIncrement {
+            val history: List<History>
+
+            if (showHistorySearchGroups) {
+                // We need to refetch all the history metadata if the offset resets back at 0
+                // in the case of a pull to refresh.
+                if (::historyGroups.isInitialized.not() || offset == 0L) {
+                    val historyMetadata =
+                        historyStorage.getHistoryMetadataSince(Long.MIN_VALUE)
+                            .filter { it.key.searchTerm != null }
+                    historyGroups = historyMetadata
+                        .groupBy { it.key.searchTerm!! }
+                        .map { (searchTerm, items) ->
+                            History.Group(
+                                id = items.first().createdAt.toInt(),
+                                title = searchTerm,
+                                visitedAt = items.first().createdAt,
```

Comment:
@grigoryk @gabrielluong Actually `getHistoryMetadataSince` returns the list ordered by `updatedAt` DESC, so the groups `visitedAt` should be `items.first().updatedAt` i.e., the group was last visited based on the most recently "visited" page in the group.

Let's fix that one before landing because it could cause the group to show up in the wrong segment? Luckily `groupBy` will retain iteration order so just switching to `updatedAt` should work.

Reply:
In guess in other words we would want display the group in the time segment based on the newest / most recent item so that would be `updatedAt`. Even if we wanted to slot the group in based on the oldest item this wouldn't work because the items are sorted by `updatedAt` not `createdAt`.